Do not discover properties resource in the classpath to avoid missing resources registration in native mode

radcortez opened the pull request:
Disable automatic discovery of configuration resources in native mode.

While it has always been enabled, it didn't have any effect because we never registered the resources in the native image. The configuration still worked, because we recorded the configuration during build time.

I did try to disable it as well for JVM mode, but it is a bit more tricky:
- Our testing API can add additional classpath resources after we build Quarkus, which wouldn't be caught with this.
- Debugging would be harder, because it would lose the the origin source from where the configuration is coming from.
- Flattening has a corner case issue described in https://github.com/quarkusio/quarkus/pull/44652#issuecomment-2497713122. This was never really an issue, but there is a test in the TCK that fails because of this.

Ideally, both JVM and native mode should be the same, but at the moment they're not. While this doesn't solve that particular issue, it should at least fix the missing registration warnings / errors without changing the current behaviour.

- Fixes #41994

Alternate proposal to https://github.com/quarkusio/quarkus/pull/42140.

radcortez committed:
Do not discover properties resource in the classpath to avoid missing resources registration in native mode

geoand commented:
CI failures seem related

radcortez commented:
> CI failures seem related

Yes

radcortez commented:
I was expecting more actually :)

zakkak commented:
@radcortez what's the status of this PR, is it ready for review and merging?

radcortez commented:
Yes, if you are happy with it :)

zakkak commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/NativeConfigBuilder.java:
```diff
@@ -0,0 +1,37 @@
+ * became a strict check during native image execution.
+ * <p>
+ * Configuration coming from classpath resources is recoded during build time in a low ordinal source, so the
+ * configuration will still be available. If the users change the ordinals of the sources in runtime, it may
+ * cause unexpected values to be returned by the config, but this has always been the case. The classpath configuration
+ * resources were never registered in the native image.
```
Is this true? 

In https://github.com/oracle/graal/discussions/10264#discussioncomment-11604091 we describe a scenario were using the classpath would work fine (assuming the user registers the resource for reflection and sets the ordinals accordingly).

radcortez replied:
With our stuff yes, we never registered the resources. The configuration is just included in a class file and we discard the resources, which may cause the issue described in https://github.com/quarkusio/quarkus/pull/44652#issuecomment-2497713122, but this has been an issue since day one and until now we never got any complaints about it.

zakkak commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/Substitutions.java:
```diff
@@ -99,4 +105,12 @@ public byte[] getClassBytes() {
+    static final class Target_AbstractLocationConfigSourceLoader {
+        @Substitute
+        protected List<ConfigSource> tryClassPath(final URI uri, final int ordinal, final ClassLoader classLoader) {
+            return Collections.emptyList();
```
If what my previous comment is right, this is a breaking (for quite a corner case I admit) change, no?

radcortez replied:
Yes, this could be a breaking change if a user has a custom source with a custom resource. If they are registering the resource manually they would need to process the file and generate some sort of representation (like us) for runtime.

Unfortunately, I don't see any good alternative. Having to forcely register all the resources is going to throw away all of our optimizations. If we leave it as is, you get the error that we are trying to prevent. (Well I guess we could add a try / catch there instead).

I still think it doesn't make sense to apply that validation to an API like `ClassLoader#getResources`. That API never fails in JVM mode, no matter what you are looking for. Granted that you can obtain different results from JVM to native if the registrations are not done properly, but the native image shouldn't try to outsmart the library or integrator by changing API behaviours.

zakkak replied:
> Unfortunately, I don't see any good alternative. 

I am personally fine with this, since it's a corner case. I am not even sure we need to document this kind of "breakage".

zakkak replied:
@radcortez I went on with adding a try/catch. Please have a look.

zakkak reviewed (dismissed):
LGTM, but I have a concern about this being a breaking change, see comments.

zakkak committed:
Catch `MissingReflectionRegistrationError` in loadConfigSources

Works around issue discussed in
https://github.com/oracle/graal/discussions/10264#discussioncomment-11604091

gsmet commented:
Hey there. What's the status of this?

zakkak commented:
> Hey there. What's the status of this?

I am waiting for @radcortez's ACK before merging.

radcortez commented:
I'd like to hold this a bit due to https://github.com/quarkusio/quarkus/issues/45436#issuecomment-2590524250, which may change a few things.

zakkak commented:
Hi @radcortez , are there any updates on this?

radcortez commented:
No, I need to pick this up again.

zakkak commented:
Hi @radcortez , is there any chance we can start working on this again?

zakkak commented:
> I guess from the Graal side, they still do not want to make any concessions about the check?

The latest update is that the missing registration exception will keep being thrown for resource accesses when using `--exact-reachability-metadata` but there is no plan to make it the default behavior (at least for the time being), while other unregistered accesses (like `getDeclaredMethods`) will start throwing missing registration exceptions by default at some point in the future.

radcortez commented:
How about `java.lang.ClassLoader#getResources`? They never replied to my comment here: https://github.com/oracle/graal/discussions/10264#discussioncomment-11586784

zakkak commented:
> How about `java.lang.ClassLoader#getResources`? They never replied to my comment here: [oracle/graal#10264 (reply in thread)](https://github.com/oracle/graal/discussions/10264#discussioncomment-11586784)

My understanding is that it will be handled just any other resource access, i.e. it will still throw an exception when `--exact-reachability-metadata` option is present but it won't throw an exception by default (at least for the time being).

zakkak commented:
@radcortez can you please also comment on what was decided in regards of this? Did you close it because it's out of date? Do you have an alternative approach in mind? Thanks

radcortez commented:
It is out of date for now. There has been a lot of changes in between, so we need to rethink a bit how to do it if required.
